File: src/client/java/com/somefrills/chestui/SaveCursorPosition.java
package com.somefrills.chestui;

import kotlin.Pair;
import net.minecraft.client.util.InputUtil;
import net.minecraft.client.MinecraftClient;

import static com.somefrills.Main.mc;

public final class SaveCursorPosition {

    private SaveCursorPosition() {}

    public static boolean active() {
        return mc.currentScreen instanceof ChestUI;
    }

    private static Pair<Double, Double> savedPositionedP1;
    private static SavedPosition savedPosition;

    public static final class SavedPosition {
        public final Pair<Double, Double> middle;
        public final Pair<Double, Double> cursor;
        public final long savedAt;

        public SavedPosition(Pair<Double, Double> middle,
                             Pair<Double, Double> cursor,
                             long savedAt) {
            this.middle = middle;
            this.cursor = cursor;
            this.savedAt = savedAt;
        }
    }

    public static void saveCursorOriginal(double x, double y) {
        if (!active()) return;

        savedPositionedP1 = new Pair<>(x, y);
    }

    public static void saveCursorMiddle(double middleX, double middleY) {
        if (!active()) return;
        if (savedPositionedP1 == null) return;

        savedPosition = new SavedPosition(
                new Pair<>(middleX, middleY),
                savedPositionedP1,
                System.currentTimeMillis()
        );
    }

    public static Pair<Double, Double> loadCursor(double middleX, double middleY) {
        if (!active()) return null;
        if (savedPosition == null) return null;

        long now = System.currentTimeMillis();

        if (now - savedPosition.savedAt > 5000) {
            savedPosition = null;
            return null;
        }

        if (Math.abs(savedPosition.middle.getFirst() - middleX) < 1
                && Math.abs(savedPosition.middle.getSecond() - middleY) < 1) {

            MinecraftClient client = MinecraftClient.getInstance();

            InputUtil.setCursorParameters(
                    client.getWindow(),
                    InputUtil.GLFW_CURSOR_NORMAL,
                    savedPosition.cursor.getFirst(),
                    savedPosition.cursor.getSecond()
            );

            return savedPosition.cursor;
        }

        return null;
    }
}

File: src/main/kotlin/com/somefrills/features/misc/SaveCursorPosition.kt
package com.somefrills.features.misc

import com.mojang.blaze3d.platform.InputConstants
import com.somefrills.Main.mc
import com.somefrills.config.FrillsMod

import com.somefrills.events.ScreenCloseEvent
import com.somefrills.events.ScreenOpenEvent
import com.somefrills.features.core.Feature
import com.somefrills.modules.FrillsFeature
import com.somefrills.features.misc.glowmob.chestui.ChestUI
import com.somefrills.events.core.EventHandle
import net.minecraft.client.Minecraft
import kotlin.math.abs

@FrillsFeature
object SaveCursorPosition : Feature(FrillsMod.config.misc.saveCursorPosition.enabled) {
    private val config get() = FrillsMod.config.misc.saveCursorPosition
    private var savedPositionedP1: Pair<Double, Double>? = null
    private var savedPosition: SavedPosition? = null

    fun active(): Boolean {
        if (config.onlyChestUI) {
            return mc.screen is ChestUI
        }
        return true
    }

    @JvmStatic
    fun saveCursorOriginal(x: Double, y: Double) {
        if (!active()) return
        savedPositionedP1 = Pair(x, y)
    }

    @JvmStatic
    fun saveCursorMiddle(middleX: Double, middleY: Double) {
        if (!active()) return

        savedPosition = SavedPosition(
            Pair(middleX, middleY),
            savedPositionedP1 ?: return,
            System.currentTimeMillis()
        )
    }

    @EventHandle
    fun onScreen(event: ScreenOpenEvent) {
        loadCursor(mc.mouseHandler.xpos(), mc.mouseHandler.ypos())
    }

    @EventHandle
    fun onScreen(event: ScreenCloseEvent) {
        savedPosition = null
    }

    @JvmStatic
    fun loadCursor(middleX: Double, middleY: Double): Pair<Double, Double>? {
        if (!active()) return null
        val pos = savedPosition ?: return null

        val now = System.currentTimeMillis()

        if (now - pos.savedAt > 5000) {
            savedPosition = null
            return null
        }

        if (abs(pos.middle.first - middleX) < 1
            && abs(pos.middle.second - middleY) < 1
        ) {
            val client = Minecraft.getInstance()

            InputConstants.grabOrReleaseMouse(
                client.window,
                InputConstants.CURSOR_NORMAL,
                pos.cursor.first,
                pos.cursor.second
            )

            return pos.cursor
        }

        return null
    }

    @JvmRecord
    data class SavedPosition(
        val middle: Pair<Double, Double>,
        val cursor: Pair<Double, Double>,
        val savedAt: Long
    )
}